### ISSUE TITLE
WCGA-AA Color contrast ratio on "screen reader only" element

### ISSUE THREAD

**Issue by sandrina-p** [ctomczyk/aslint#2]
On [my website](http://sandrina-p.net) I have several elements with the class `.sr-only` that is used to hide visually the element and shows it only for screen readers.

The issue is aslint says it needs contrast.

```
Expected color contrast ratio at least 4.5:1, but got ratio: 4.26:1, background-color: #rgb(242, 242, 242), color: #rgb(0, 128, 128), level: AA, size: small.
Node: <h2 class="sr-only">Chat Bot sections</h2>
XPath: XPath string 
/html/body/main/div/nav/h2  
Impact: major
```

That happens because if I remove `sr-only` you can see the element has a teal color.

This is the CSS for `.sr-only`
```css
.sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0,0,0,0);
    border: 0;
}
```

I think your tool could be able to detect `.sr-only` elements and ignore them.

**Comment by ctomczyk** [comment]
The issue is because `color-contrast-aa` and `color-contrast-aaa` rule is checking for element visibility and that checking wasn't passing correctly through the case where element is visually hidden, but exposed to assistive technologies. It's should be fixed now in ASLint version 0.0.699.

**Comment by ctomczyk** [comment]
@sandrina-p Try to test how it works now.

**Comment by sandrina-p** [comment]
Well, it doesn't give the error, but now all the contrast AA errors disappeared. 😅

**Comment by ctomczyk** [comment]
When you land on [your page](http://sandrina-p.net/) then only few elements are visible at that moment. The rest elements are simply hidden unless the user starts scrolling down. That's why contrast test can't be done because elements are invisible.

However, if you scroll down to the very bottom and reevaluate all tests (a button on right side to Focus checkbox) then you'll find contrast ratio tests results.

The middle content seems to be fine in terms of contrast ratio level AA.

**Comment by sandrina-p** [comment]
Humm not really, I discovered with ASLint I had contrast ratio level AA issues.
Background: `#f2f2f2` with text `#008080` (navigation links for example) fails level AA issues with, as [you can confirm here](http://webaim.org/resources/contrastchecker/)

But now the ASLint doesn't shows the error...

**Comment by ctomczyk** [comment]
I'll look into it more closely yet to find out what's causing the issue. Thanks for reporting it.

**Comment by sandrina-p** [comment]
Okay! You are welcome!

**Comment by ctomczyk** [comment]
It seems that I can't determine in a reliable way if element is invisible. There are too many combinations:

- it can be hidden, but available for assistive technologies and there are multiple different technique used
- it can have `opacity: 0`
- it can have `display: none`
- it can be temporary invisible, e.g. due to lazy loading
- it can have `text-indent` with a big value
- it can have `width: 0` and `height: 0`

What I've done is to determine if element is possibly hidden, but do test for color contrast and add note about hide state.

Would you mind to check how it works now? The version of ASLint should be 0.0.718 at the moment of writing this comment.

**Comment by sandrina-p** [comment]
Yah, now it shows the error even if it's hidden.

You could show the `Note` about visibility only if you think it is hidden. For instance, nothing on my Nav Links says they are hidden but they have the note:
> Note: The element is possibly visually hidden and that state might be temporary. For example the lazy loading technique can be used and the image will be visible at some point.

(To test the warning changed the navbar links color to a lighter gray on DevTools to trigger Level AA warning)

**Comment by ctomczyk** [comment]
That's what I do. I mean, I do a test if element is possibly hidden using:

    function isElementVisible(element) {

        var style,
            centerPosition,
            elementFromPoint,
            textIndent,
            REASONABLE_TEXT_INDENT = -100;

        if (object.isHtmlElement(element) === false) {
            return false;
        }

        style = global.getComputedStyle(element, null);

        if (style.display === 'none' || style.visibility !== 'visible' || style.opacity < 0.1 || element.offsetWidth + element.offsetHeight + element.getBoundingClientRect().height + element.getBoundingClientRect().width === 0) {
            return false;
        }

        centerPosition = {
            x: element.getBoundingClientRect().left + element.offsetWidth / 2,
            y: element.getBoundingClientRect().top + element.offsetHeight / 2
        };

        if (centerPosition.x < 0 || centerPosition.x > (document.documentElement.clientWidth || window.innerWidth) || centerPosition.y < 0 || centerPosition.y > (document.documentElement.clientHeight || window.innerHeight)) {
            return false;
        }

        textIndent = parseInt(style.textIndent, 10);

        elementFromPoint = document.elementFromPoint(centerPosition.x, centerPosition.y);

        while (elementFromPoint) {
            if (elementFromPoint === element && textIndent > REASONABLE_TEXT_INDENT) {
                return true;
            }
            elementFromPoint = elementFromPoint.parentNode;
        }

        return false;
    }

and if the result is `true` then I add the note.

**Note**: `global` refers to `window` here.

I will look into Nav Links more closely. I've improved a bit rules `color-contrast-aa` and `color-contrast-aaa`. Version 0.0.723 has been deployed.

**Comment by ctomczyk** [comment]
It should work correctly now with version 0.0.799. You have no more color contrast issues on level AA.
